recipe/2024-01-global-scope-secret-store.md

```diff
@@ -0,0 +1,261 @@
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
+   recipeConfig: {
+      terraform: {
+        authentication:{
+          git:{
+            pat: {
+              "dev.azure.com": {
++               secret: secretStoreAzure.id
+              }
+              "github.com": {
++                secret: secretStoreGithub.id
+              }
+            }
+          }
+        }
+     }
+   }
+    recipes: {      
+      'Applications.Datastores/mongoDatabases':{
+        default: {
+          templateKind: 'terraform'
+          templatePath: 'https://dev.azure.com/test-private-repo'
+        }
+      }
+    }
+  }
+}
+```
+
+
+## Design
+Today we use Application.Core/secretstores resource to store secrets/credentials information for an application but we cannot use it in this scenario as secretStores is application scoped and expects to have an application created before we create a secretStore.
+
+And we cannot have secretStore created for environment scope in this scenario as we are adding the secret to the environment which creates cyclic dependency between environment and secretStore resource.
+
+#### Application.Core/secretStores as a global scoped resource.
+We could make secretStores as a global scoped resource by removing application as a required property.
+
+```diff
+"SecretStoreProperties": {
+      "type": "object",
+      "description": "The properties of SecretStore",
+      "properties": {
+        "environment": {
+          ...
+        },
+        "application": {
+          ...
+        },
+        "provisioningState": {
+          ...
+        },
+        "status": {
+          ...
+        },
+        "type": {
+          ...
+        },
+        "data": {
+          "type": "object",
+          "description": "An object to represent key-value type secrets",
+          "additionalProperties": {
+            "$ref": "#/definitions/SecretValueProperties"
+          }
+        },
+        "resource": {
+          "type": "string",
+          "description": "The resource id of external secret store."
+        }
+      },
+      "required": [
+-        "application",
+        "data"
+      ]
+    },
+```
+With this change we can create a secret store resource before an application or environment is created. But, if the user is creating a secretStore of kind kubernetes secret, it is deployed in application/environment namespace. But for secret resource with global scope use the `resource` property which is used to specify the secret ref of the existing secret to provide namespace and secret name details. User is expected to provide these details in `<namespace>/<secretName>` format for secretStore with global scope.
+
+```diff
+resource secretStore 'Applications.Core/secretStores@2023-10-01-preview' = {
+  name: 'github'
+  properties:{
+-  	app: app.id
++   resource: <namespace>/<secretName>
```

Comment:
@vishwahiremat could you confirm if it was decided to go with `<namespace>/<secretName>` or something more generic that would work with other resource types as well as mentioned in this comment https://github.com/radius-project/design-notes/pull/38/files#r1466963848? Would be great if you could add a quick note in the design on how this would extend for other resource types in the future.

Reply:
we decided to go with <namespace>/<secretname> format. and I will update it